### PR TITLE
Parallelise KNN pursuit over pairs of both arrays' chunks

### PR DESCRIPTION
**Pull request by jjerphan**
**Main changes**:
 - remove single chunking strategy
 - reverted from working memory to chunk size (number of vector per chunk)
  - rename X and Y to X_train and X_test and permute them on the APIs
 - work on two double chunking implementations with different parallelization strategies:
     -   `_parallel_knn_on_X_test` computes KNN on parallel chunks of X_test. It comes with less overhead (there's no need for synchronisation) but is potentially less suitable in practice.
     -   `_parallel_knn_on_X_train` computes KNN on parallel chunks of X_train. It is potentially more suitable but comes with more overhead (we need extra data-strucutres for each threads and need to synchronize heaps, see `with gil` context-manager section) 
  - adapt benchmark setup accordingly

---

Comes after #3.
Comes before #7.

### CODE REVIEW

**Comment by jjerphan** [comment]
Benchmarks results as of 85b76ca3d1:

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -40,10 +40,8 @@ jobs:
         run: |
           export NUM_CORES=`python -c "import joblib; print(joblib.cpu_count())"`
           export OMP_NUM_THREADS=$NUM_CORES
-          # Avoid threads over-subscription in parallel implementation
-          # TODO: remove once threadpoolctl is set.
-          export MKL_NUM_THREADS=1
-          export OPENBLAS_NUM_THREADS=1
+          export MKL_NUM_THREADS=$NUM_CORES
+          export OPENBLAS_NUM_THREADS=$NUM_CORES
```
Comment (ogrisel) [inline .github/workflows/benchmark.yml]: Now that you use threadpoolctl you should be amble to not set those variables anymore.

File: benchmarks/benchmark.py [hunk 2]
```diff
@@ -5,70 +5,76 @@
...
+chunk_sizes = config["chunk_size"]
 n_neighbors = config["n_neighbors"]
 estimators = config["estimators"]
 
 N_TRIALS = 5
 one_GiB = 1e9
 benchmarks = pd.DataFrame()
 
+# Avoid thread over-subscription by BLAS
+with threadpool_limits(limits=1, user_api="blas"):
```
Comment (ogrisel) [inline benchmarks/benchmark.py]: Could you set this in the code of the `kneighbors` function instead?

**Comment by ogrisel** [comment]
Could you also add the size in MB of the temporary datastructures to the x axis of the plot to be able to interpret those in terms of cache sizes?

It would also be interesting to do a sequential run for all libraries by setting all the `_NUM_THREADS` variable to 1.

**Comment by ogrisel** [comment]
And maybe add even smaller chunk sizes suchs as 256 and 512.

**Comment by jjerphan** [comment]
> Could you also add the size in MB of the temporary datastructures to the x axis of the plot to be able to interpret those in terms of cache sizes?

I am doing an analysis on the new implementation taking this into account.

I probably will report some kind heatmap matrix with color indicating average runtime against n and k, and values for estimating working memories based on data-structures considered (this last step can be done easily _a posteriori_).

**Comment by jjerphan** [comment]
Benchmarks results as of 4fa804d:
 - [Parallel execution](https://github.com/scikit-learn-inria-fondation/pdist_aggregation/files/6524590/par_exec_results_4fa804dde3.pdf) (`*_NUM_THREADS` were not set)
 - [Sequential execution](https://github.com/scikit-learn-inria-fondation/pdist_aggregation/files/6539850/4fa804dd_sequential_results.pdf) (`*_NUM_THREADS=1`, and forcing only one core to be used via `taskset(1)`)
 

Edit: updated pdf file for sequential execution

**Comment by jjerphan** [comment]
[Simple throughput estimation @ `4fa804d`](https://github.com/scikit-learn-inria-fondation/pdist_aggregation/files/6546117/throughput_4fa804dde3.pdf)  (aggregation of results given above).

**Pull request merged**